### PR TITLE
Add validation in Governor on ERC-721 or ERC-1155 received

### LINKED ISSUE #4307: Governor should revert on ERC1155 and ERC721 receive hooks when it's not executor
## Description

Currently, the Governor implementation uses `onERC1155Received`, `onERC1155BatchReceived`, and `onERC721Received` to acknowledge token receives. However, if the Governor is not the executor itself, it can't operate on those tokens except using the `relay` function.

We consider that all of these receive hooks can also validate that the Governor is also the executor, as in the `receive()` function:

```solidity
/**
 * @dev Function to receive ETH that will be handled by the governor (disabled if executor is a third party contract)
 */
receive() external payable virtual {
    if (_executor() != address(this)) {
        revert GovernorDepositDisabled(address(this));
    }
}
```

### PR DESCRIPTION
**Pull request by clauBv23**
Fixes #4307  

Add validation to `onERC1155Received`, `onERC1155BatchReceived`, and `onERC721Received` hooks in Governor to check it is the executor.
Add test to the validations.

Note: This PR should be refactored after merging #4261 and #4284 

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [ ] Changeset entry (run `npx changeset add`)

### CODE REVIEW

**Review by Amxx — COMMENTED** [review]
If possible, the expectRevert should not be unspecified.

File: contracts/governance/Governor.sol [hunk 1]
```diff
@@ -603,14 +603,15 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor, IERC721Receive
     }
 
     /**
-     * @dev See {IERC721Receiver-onERC721Received}.
+     * @dev See {IERC721Receiver-onERC721Received}(disabled if executor is a third party contract).
```
Comment (ernestognw) [inline contracts/governance/Governor.sol:606]:
```suggestion
     * @dev See {IERC721Receiver-onERC721Received}.
     * Receiving tokens is disabled if the governance executor is other than the governor itself (eg. when using with a timelock).
```

File: contracts/governance/Governor.sol [hunk 2]
```diff
@@ -603,14 +603,15 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor, IERC721Receive
...
      */
     function onERC721Received(address, address, uint256, bytes memory) public virtual override returns (bytes4) {
+        require(_executor() == address(this), "Governor: must send to executor");
         return this.onERC721Received.selector;
     }
 
     /**
-     * @dev See {IERC1155Receiver-onERC1155Received}.
+     * @dev See {IERC1155Receiver-onERC1155Received}(disabled if executor is a third party contract).
```
Comment (ernestognw) [inline contracts/governance/Governor.sol:614]:
```suggestion
     * @dev See {IERC721Receiver-onERC721Received}.
     * Receiving tokens is disabled if the governance executor is other than the governor itself (eg. when using with a timelock).
```

File: contracts/governance/Governor.sol [hunk 3]
```diff
@@ -619,11 +620,12 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor, IERC721Receive
         uint256,
         bytes memory
     ) public virtual override returns (bytes4) {
+        require(_executor() == address(this), "Governor: must send to executor");
         return this.onERC1155Received.selector;
     }
 
     /**
-     * @dev See {IERC1155Receiver-onERC1155BatchReceived}.
+     * @dev See {IERC1155Receiver-onERC1155BatchReceived}(disabled if executor is a third party contract).
```
Comment (ernestognw) [inline contracts/governance/Governor.sol:628]:
```suggestion
     * @dev See {IERC721Receiver-onERC721Received}.
     * Receiving tokens is disabled if the governance executor is other than the governor itself (eg. when using with a timelock).
```

**Review by ernestognw — DISMISSED** [review]
This looks good @clauBv23! Thanks. I'm adding a few suggestions and also requesting review from others.

Thanks!

File: contracts/governance/Governor.sol (hunk 2 above)
Comment (clauBv23) [reply contracts/governance/Governor.sol:614]: Hi @ernestognw, thanks for the suggestions!

Maybe is confusion or I'm understanding something wrong but, if is the `onERC1155Received` hook why 

 ``` solidity 
     * @dev See {IERC1155Receiver-onERC1155Received}.
```
should be changed by

``` solidity 
    * @dev See {IERC721Receiver-onERC721Received}.
```

Comment (Amxx) [reply contracts/governance/Governor.sol:614]: Should be 
```suggestion
     * @dev See {IERC1155Receiver-onERC1155Received}.
     * Receiving tokens is disabled if the governance executor is other than the governor itself (eg. when using with a timelock).
```

File: contracts/governance/Governor.sol (hunk 3 above)
Comment (Amxx) [reply contracts/governance/Governor.sol:628]: same here, should be 
```suggestion
     * @dev See {IERC1155Receiver-onERC1155BatchReceived}
     * Receiving tokens is disabled if the governance executor is other than the governor itself (eg. when using with a timelock).
```

File: contracts/governance/Governor.sol (hunk 2 above)
Comment (ernestognw) [reply contracts/governance/Governor.sol]: Completely my bad. Yes, it has to be as @Amxx suggests. Same for others.

**Review by Amxx — APPROVED** [review]
LGTM

**Comment by Amxx** [comment]
@frangio @ernestognw I merged master into this and realised the following:

ERC1155 acceptance hook was written in a way that bubble error string, but not custom error. The result is that the Governor custom error (`GovernorDisabledDeposit`) is replaced by the default one (`ERC1155InvalidReceiver`)

Do we want to change the ERC1155 hook to use the syntax from ERC721 (which bubble the custom errors)?

**Comment by frangio** [comment]
> Do we want to change the ERC1155 hook to use the syntax from ERC721 (which bubble the custom errors)?

I'd say yes.

File: test/token/ERC1155/ERC1155.behavior.js [hunk 4]
```diff
@@ -385,23 +392,55 @@ function shouldBehaveLikeERC1155([minter, firstTokenHolder, secondTokenHolder, m
...
+
+        it('with custom error', async function () {
+          const receiver = await ERC1155ReceiverMock.new(
+            RECEIVER_SINGLE_MAGIC_VALUE,
+            RevertType.Custom,
+            RECEIVER_BATCH_MAGIC_VALUE,
+            RevertType.None,
+          );
+
+          await expectRevert(
```
Comment (clauBv23) [inline test/token/ERC1155/ERC1155.behavior.js:436]:
```suggestion
          await expectRevertCustomError(
```

**Review by frangio — DISMISSED** [review]
Thanks!

File: .changeset/thin-camels-matter.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+'openzeppelin-solidity': minor
+---
+
+`ERC1155`: Bubble custom errors triggered in the `onERC1155Received` and `onERC1155BatchReceived` hooks.
```
Comment (ernestognw) [inline .changeset/thin-camels-matter.md:5]: Nit, but it's not bubbling just custom errors

```suggestion
`ERC1155`: Bubble errors triggered in the `onERC1155Received` and `onERC1155BatchReceived` hooks.
```

**Review by ernestognw — DISMISSED** [review]
Thanks!

Small nit but I don't want to block. I'll let @frangio decide since he can merge even if the approval is dismissed

**Review by frangio — APPROVED** [review]

**Pull request merged**